### PR TITLE
Replace String.IndexOf() == -1 by !String.Contains

### PR DESCRIPTION
**Pull request by pgovind**
Set of changes triggered by dotnet/roslyn-analyzers#3534. These are all generated by the code-fix.

### CODE REVIEW

**Comment by pgovind** [comment]
2 errors still exist locally on my machine:

Microsoft\Win32\RegistryKey.cs(633,17): error CA1834: Use 'String.Contains' instead of 'String.IndexOf' to check for membership [D:\repos\runtime\src\libraries\Microsoft.Win32.Registry\src\Microsoft.Win32.Registry.csproj]
System\ServiceProcess\ServiceController.cs(428,57): error CA1834: Use 'String.Contains' instead of 'String.IndexOf' to check for membership [D:\repos\runtime\src\libraries\System.ServiceProcess.ServiceController\src\System.ServiceProcess.ServiceController.csproj]

Question for @mavasani: These should not have been tagged by dotnet/roslyn-analyzers#3534. Both the projects above have this for their TFM:

```
<TargetFrameworks>$(NetCoreAppCurrent)-Windows_NT;netstandard2.0;netstandard2.0-Windows_NT;net461-Windows_NT;$(NetFrameworkCurrent)-Windows_NT</TargetFrameworks>
```

What TFM does the analyzer run on in VS? I'm guessing it's `$(NetFrameworkCurrent)-Windows_NT`? The problem is that `string.Contains(string, StringComparison)` does not exist in netstandard2.0 but it does in netcoreapp5(and the latest netframework). I've guarded against it here: https://github.com/dotnet/roslyn-analyzers/pull/3534/files#diff-1dedda5e67da73be9cc5ae49b231ab91R69 but I don't think that works when targeting multiple TFMs. How do I fix this?

**Comment by MihaZupan** [comment]
Why is `IndexOf(char)` getting replaced by `Contains(char, StringComparison.CurrentCulture)` if `Contains(char)` is the one with the same behavior.

**Comment by pgovind** [comment]
> Why is IndexOf(char) getting replaced by Contains(char, StringComparison.CurrentCulture) if Contains(char) is the one with the same behavior.

That's because `string.IndexOf` and `string.Contains` differ in their default comparison arguments. See https://source.dot.net/#System.Private.CoreLib/String.Searching.cs,301 and https://github.com/dotnet/runtime/issues/33785#issuecomment-602248589

**Comment by mavasani** [comment]
> What TFM does the analyzer run on in VS? 

Analyzers are TFM agnostic, they just run on whichever TFM the active file/project context targets.

>  I don't think that works when targeting multiple TFMs. How do I fix this?

It is a known behavior that any analyzer might report issues on a piece of code in multi-TFM project for subset of the TFM contexts, but not on all contexts, especially if the API that is being analyzed is not present on all TFMs. Analyzer implementation cannot and should not handle these cases. The end user code consuming the analyzer should address this in either of the following ways:
1. add conditional pre-processor directives based on TFM to choose appropriate APIs to invoke for each TFM and fix the violation in flagged contexts OR
2. suppress such analyzer violations in source, likely with a pragma

**Comment by MihaZupan** [comment]
> That's because string.IndexOf and string.Contains differ in their default comparison arguments. See https://source.dot.net/#System.Private.CoreLib/String.Searching.cs,301 and #33785 (comment)

For strings, yes. But the `char` overloads of Contains and IndexOf do ordinal searches.

Going from `IndexOf(char)` to `Contains(char, CurrentCulture)` degrades perf.

**Comment by pgovind** [comment]
> For strings, yes. But the char overloads of Contains and IndexOf do ordinal searches.

Yup, you're right. I missed that. I'll update the PR

File: src/libraries/Common/src/System/Drawing/ColorConverterCommon.cs [hunk 1]
```diff
@@ -37,7 +37,7 @@ public static Color ConvertFromString(string strValue, CultureInfo culture)
             // If the value is a 6 digit hex number only, then
             // we want to treat the Alpha as 255, not 0
             //
-            if (text.IndexOf(sep) == -1)
+            if (!text.Contains(sep, StringComparison.Ordinal))
```
Comment (alaatm) [inline src/libraries/Common/src/System/Drawing/ColorConverterCommon.cs:40]: The correct replacement of `IndexOf(char)` should be `Contains(char)`. `Contains(char, StringComparison.Ordinal)` will just call `IndexOf(char)` which is what you want replaced.

File: src/libraries/System.DirectoryServices.AccountManagement/src/System/DirectoryServices/AccountManagement/Context.cs [hunk 2]
```diff
@@ -104,8 +104,7 @@ private bool BindSam(string target, string userName, string password)
                 // if they just passed user then append the machine name here.
                 if (null != userName)
                 {
-                    int index = userName.IndexOf('\\');
-                    if (index == -1)
+                    if (!userName.Contains('\\'))
```
Comment (GrabYourPitchforks) [inline src/libraries/System.DirectoryServices.AccountManagement/src/System/DirectoryServices/AccountManagement/Context.cs:107]: This DLL is compiled against netstandard2.0 and netfx, which don't contain `string.Contains(char)`. This will lead to compilation failures.

**Comment by danmosemsft** [comment]
> Analyzer implementation cannot and should not handle these cases. The end user code consuming the analyzer should address this in either of the following ways:

Oftentimes the developer will want to just not apply the fix, rather than introducing conditional compilation or suppression. Why can the analyzer not be told (or find out) that the the fix will cause a build error?

**Comment by mavasani** [comment]
@danmosemsft Analyzers run against a single active project context or compilation, within a single csc invocation. It has no knowledge of concepts like a project, TFMs, etc.

**Comment by mavasani** [comment]
Actually, putting more thought into, let me rephrase. _Current_ analyzer configuration support does not allow custom MSBuild properties to be passed down to the analyzer. However, @chsienki @jaredpar are working on a feature request https://github.com/dotnet/roslyn/issues/42219 that will allow custom MSBuild properties to be passed into analyzers as analyzer options. Target TFMs and active TFM seem like very useful MSBuild properties to be passed to analyzers, and once they are available, we can update the analyzer to bail out if any one of the TFMs does not support the API it analyzes.

File: src/libraries/System.Net.WebProxy/src/System/Net/WebProxy.cs [hunk 3]
```diff
@@ -95,8 +95,7 @@ public bool UseDefaultCredentials
         }
 
         private static Uri? CreateProxyUri(string? address) =>
-            address == null ? null :
-            address.IndexOf("://", StringComparison.Ordinal) == -1 ? new Uri("http://" + address) :
+            address == null ? null : !address.Contains("://", StringComparison.Ordinal) ? new Uri("http://" + address) :
```
Comment (stephentoub) [inline src/libraries/System.Net.WebProxy/src/System/Net/WebProxy.cs:98]: The default for `string.Contains(string)` is ordinal, right?  So the second arg here shouldn't be necessary.

**Comment by mavasani** [comment]
Seems like @chsienki already has a PR for adding support for passing MSBuild properties to analyzers: https://github.com/dotnet/roslyn/pull/43617

**Comment by danmosemsft** [comment]
> once they are available, we can update the analyzer to bail out if any one of the TFMs does not support the API it analyzes.

I'm assuming that the key knowledge the analyzer wants is what API is available (rather than, say, what runtime features exist). For that purpose I'm guessing knowing the TFM isn't super helpful (analyzers can't reasonably encode knowledge of which API are available in every TFM past and future). What it wants to know is: what API is available.

**Comment by danmosemsft** [comment]
What about `IndexOf(...) >= 0`? There seem to be quite a few cases, eg

https://github.com/dotnet/runtime/blob/4f9ae42d861fcb4be2fcd5d3d55d5f227d30e723/src/libraries/System.Private.Xml/src/System/Xml/Core/XmlCharCheckingWriterAsync.cs#L104

In customer codebases it may be more common, depending on style (since it's obviously equivalent to != -1)

**Comment by mavasani** [comment]
> I'm assuming that the key knowledge the analyzer wants is what API is available (rather than, say, what runtime features exist). For that purpose I'm guessing knowing the TFM isn't super helpful (analyzers can't reasonably encode knowledge of which API are available in every TFM past and future). What it wants to know is: what API is available.

In that case, I don't think this is possible within an analyzer. Analyzers only run against a single compilation in a single csc invocation, and has no way to know if the API available in current compilation will be available or not in all compilations for the respective TFMs that the project targets.

**Comment by danmosemsft** [comment]
> In that case, I don't think this is possible within an analyzer. Analyzers only run against a single compilation in a single csc invocation, and has no way to know if the API available in current compilation will be available or not in all compilations for the respective TFMs that the project targets.

I guess it need not be perfect - it could hard code a few TFM's that are commonly seen and definitely don't have the API, just to reduce the noise. Anyway -- this is just an aside. Thanks.

File: src/libraries/System.Net.WebProxy/src/System/Net/WebProxy.cs (hunk 3 above)
Comment (pgovind) [reply src/libraries/System.Net.WebProxy/src/System/Net/WebProxy.cs:98]: Yup, it isn't necessary. This is a bug in the analyzer. I realized that the `Contains(string, StringComparison)` method calls through to the `IndexOf(string, StringComparison)` only after I put this PR up. I need to reduce the number of cases the analyzer tags

**Comment by pgovind** [comment]
Closing in favor of another PR that I'll soon open
